### ISSUE TITLE
Weird Regression

### ISSUE THREAD

**Issue by gnat42** [scheb/2fa#119]
**Bundle version**: 5.13.1
**Symfony version**: 5.4.1
**PHP version**: 7.4.27
**Using authenticators**: YES
**Description**

We had a regression where we no longer get the 2fa login screen. We are directed to the normal landing page - however its messed up because the token is the TwoFactorToken and so there are no roles so our menu etc are all messed up. To be clear, 2FA is optional on a per user or firm basis. I narrowed down the breaking change in our code which seems super simple and not something that should break it.

**Additional Context**

Here's our security.yaml

```
security:
    enable_authenticator_manager: true

    password_hashers:
        legacy:
            algorithm: sha256

        NS\CoreDomain\Firm\User:
            algorithm: auto
            migrate_from:
                - legacy
        NS\AdminBundle\Entity\User:
            algorithm: bcrypt
            cost: 14

    role_hierarchy:
        ROLE_3RD_PARTY:    ROLE_USER
        ROLE_UNVERIFIED:   ROLE_USER
        ROLE_MANAGEMENT:   ROLE_USER
        ROLE_STAFF:   ROLE_MANAGEMENT
        ROLE_ADMIN:   ROLE_MANAGEMENT
        ROLE_PROLEGIS_SUPER_ADMIN:  [ ROLE_USER, ROLE_PROLEGIS_ADMIN, ROLE_ALLOWED_TO_SWITCH, ROLE_PROLEGIS_ADMIN_CLIENTS, ROLE_PROLEGIS_ADMIN_REAL_ESTATE, ROLE_PROLEGIS_ADMIN_PRECEDENTS ]

    providers:
        users:
            entity: { class: NS\CoreDomain\Firm\User }
        admin:
            entity: { class: NS\AdminBundle\Entity\User, property: email }
    firewalls:
        dev:
            pattern:  ^/(_(profiler|wdt)|css|images|js)/
            security: false

        admin:
            pattern: ^/admin
            provider: admin
            form_login:
                login_path: /admin/login
                check_path: /admin/login_check
                default_target_path: /admin/dashboard
                always_use_default_target_path: true
            logout:
                path: /admin/logout
            two_factor:
                auth_form_path: admin_2fa_login
                check_path: admin_2fa_login_check
                default_target_path: /admin/dashboard
                always_use_default_target_path: true
            entry_point: form_login

        default:
            pattern: ^/
            provider: users
            user_checker: NS\SiteBundle\User\ActiveUserChecker
            custom_authenticators:
                - NS\PracticeBundle\Authenticators\Api\ApiAuthenticator
                - NS\PracticeBundle\Authenticators\ThirdParty\ThirdPartyUserAuthenticator
                - NS\PracticeBundle\Authenticators\Token\TokenExchangeAuthenticator
            form_login:
                login_path: /login
                check_path: /login_check
                default_target_path: /practice/dashboard
                always_use_default_target_path: true
            logout:
                path: /logout
            two_factor:
                auth_form_path: 2fa_login
                check_path: 2fa_login_check
                default_target_path: /practice/dashboard
                always_use_default_target_path: true
            entry_point: form_login

    access_control:
        # This makes the logout route available during two-factor authentication, allows the user to cancel
        - { path: ^/admin/logout, role: PUBLIC_ACCESS }
        - { path: ^/logout, role: PUBLIC_ACCESS }
        # This ensures that the form can only be accessed when two-factor authentication is in progress
        - { path: ^/admin/2fa, role: IS_AUTHENTICATED_2FA_IN_PROGRESS }
        - { path: ^/2fa, role: IS_AUTHENTICATED_2FA_IN_PROGRESS }
        - { path: ^/login, roles: PUBLIC_ACCESS }
        - { path: ^/login_check, roles: [ ROLE_UNVERIFIED ] }
        - { path: ^/practice/dashboard, roles: [ ROLE_UNVERIFIED, ROLE_USER ] }
        - { path: ^/, roles: PUBLIC_ACCESS }
```

I've condensed it a bit. to only include relevant access_control rules. The above configuration works. When I change this:

```
        - { path: ^/practice/dashboard, roles: [ ROLE_UNVERIFIED, ROLE_USER ] }

to 
        - { path: ^/practice/dashboard, roles: [ ROLE_UNVERIFIED, ROLE_LAW, ROLE_INTERNAL ] }
```

Then when we login we are redirected to /practice/dashboard after login, but with no roles there is no menu etc. I haven't been able to dig into why that is yet. I had tried xdebug a bunch to isolate the issue but haven't found the problem. Then bisected a few hundred commits to find this one commit that changed a few things in security.yaml but with trial and error realized that the breakage is because of that one line change.

To explain our roles, role_unverified is assigned a newly registered user that hasn't clicked on the verification email. then we recently added a few different types of accounts and ROLE_LAW and ROLE_INTERNAL are mutually exclusive. You will only have one or the other of them - not sure if that's part of the problem.

/practice/dashboard is where you are redirected to on successful login.

Any help or where to look would be greatly appreciated. I did read the troubleshooting  and according got https://symfony.com/bundles/SchebTwoFactorBundle/current/troubleshooting.html#troubleshooting step 2 this is an unknown error.

**Comment by scheb** [comment]
```
- { path: ^/, roles: PUBLIC_ACCESS }
```
This looks suspicious. It makes every other page accessible during 2fa. My guess: because of that rule there's no need for the bundle to redirect to the 2fa authentication form.

You should be able to safely remove that rule. Then all those page should still be accessible without authentication, but they aren't explicitly whitelisted to be accessible "any time". It's a slight detail, but for access control in the 2fa state it does matter.

**Comment by gnat42** [comment]
I gave that a try but it doesn't fix the issue. Its the last one in the list and from what I know Symfony will go from top to bottom and first defined is matched. This rule is there for the last bit of the system that has a bunch of public pages, and has always been like that. 100% I can reproduce the working and non-working versions with:

```diff
-        - { path: ^/practice/dashboard, roles: [ ROLE_UNVERIFIED, ROLE_USER ] } 
+        - { path: ^/practice/dashboard, roles: [ ROLE_UNVERIFIED, ROLE_LAW, ROLE_INTERNAL ] }
        
```

Also thanks for the very quick response. Appreciated!

**Comment by gnat42** [comment]
I could probably dig into this a bit more if I understood which services intercept the login and redirect to the 2FA. I'm not entirely sure where/how that happens so don't know what to step through with the two situations differently to see where it gets messed up. What I also find odd is that Symfony redirects me to a page that with zero roles I should get an access denied...

**Comment by gnat42** [comment]
Ok, one more update. I've been tracing code execution and have found the difference but not the reason why yet.

https://github.com/scheb/2fa/blob/6.x/src/bundle/Security/Authorization/TwoFactorAccessDecider.php#L49 

The working case doesn't return true on line 50 there, the non-working case does. So 

https://github.com/scheb/2fa/blob/6.x/src/bundle/Security/Http/Firewall/TwoFactorAccessListener.php#L57

Doesn't thrown an access denied exception.

I'll continue digging.

**Comment by scheb** [comment]
`$this->accessDecisionManager` is Symfony's access control mechanism. So there must be some voter in the system, that votes "access granted" for one of the `$attributes`.

If you check that request with the profiler, go to the "Security" tab, and then "Access Decision", you can see the access checks performed. Use that (no so obvious) "Show voter details" link to unveil the individual voters and how they voted. There must be one that voted "access granted" and that's to look into further.

**Comment by gnat42** [comment]
Hello @scheb ,

  Thank you very much for pointing me at the security tab of the debug bar. I've used it for so many things but wasn't aware of that particular tab. It was much faster to find which voter was at fault there than xdebug would have been!!

  I found the voter was one we had that was passing when it should have been denying entry for a particular case.

I hope you have a great new year and thanks for your assistance!
